### PR TITLE
Optimize `gp_*_cov` functions

### PR DESCRIPTION
**Pull request by t4c1**
## Summary

`gp_*_cov` functions can be optimized by reordering the work they do, so that memory is accessed in blocks. This way the data in CPU cache can be reused multiple times and fewer transfers between CPU and RAM are needed.

Optimal value of block size depends on the amount of cache available. I experimentally optimized it for my CPU. Although, it is not very sensitive, so the value should be fine on other CPUs as well.

I also reimplemented rev overloads using `reverse_pass_callback`, reducing the amount of code.

I did not benchmark all of the modified functions, but all I did benchmark were faster. The largest speedup I observed was 90%.

## Tests
This is just a refactor. No new tests.

## Side Effects
None.

## Release notes
`gp_*_cov` functions are faster, especially for large amount of data.

## Checklist

- [ ] Math issue #(issue number)

- [ ] Copyright holder: Tadej Ciglarič

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [ ] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [ ] the code is written in idiomatic C++ and changes are documented in the doxygen

- [ ] the new changes are tested

### CODE REVIEW

File: stan/math/prim/fun/gp_dot_prod_cov.hpp [hunk 1]
```diff
@@ -54,15 +54,21 @@ gp_dot_prod_cov(const std::vector<Eigen::Matrix<T_x, Eigen::Dynamic, 1>> &x,
...
+      for (size_t j = jb; j < std::min(x_size, jb + block_size); ++j) {
+        cov.coeffRef(j, j) = sigma_sq + dot_self(x[j]);
+        for (size_t i = std::max(ib, j + 1);
+             i < std::min(x_size, ib + block_size); ++i) {
+          cov.coeffRef(j, i) = cov.coeffRef(i, j)
+              = sigma_sq + dot_product(x[i], x[j]);
+        }
+      }
     }
   }
```
Comment (SteveBronder) [inline stan/math/prim/fun/gp_dot_prod_cov.hpp:70]: How does the block schema work here wrt cache friendlyness? I'm surprised most of the time isn't being spent in `dot_product(x[i], x[j])`. I wonder how much going off of fixed block sizes is actually helping compiler loop optimizations.

**Comment by SteveBronder** [comment]
Just wanted to check and wrote a test you can add to mix

```cpp
#include <test/unit/math/test_ad.hpp>
#include <vector>

/**
 supported signatures in stanc3
 gp_dot_prod_cov(array[] real, real) => matrix
 gp_dot_prod_cov(array[] real, array[] real, real) => matrix
 gp_dot_prod_cov(array[] real, array[] real, real) => matrix
 gp_dot_prod_cov(array[] vector, real) => matrix
 gp_dot_prod_cov(array[] vector, array[] vector, real) => matrix
*/
TEST(mathMixMatFun, gp_dot_prod_cov) {
  auto f = [](const auto& x, const auto& sigma) {
    return stan::math::gp_dot_prod_cov(x, sigma);
  };

  auto ff = [](const auto& x1, const auto& x2, const auto& sigma) {
    return stan::math::gp_dot_prod_cov(x1, x2, sigma);
  };

  double sigma = 0.2;
  std::vector<double> x_stdreal{-2, -1, -0.5};
  stan::test::expect_ad(f, x_stdreal, sigma);
  stan::test::expect_ad(ff, x_stdreal, x_stdreal, sigma);
  Eigen::VectorXd x_vec = Eigen::VectorXd::Random(2);
  std::vector<Eigen::VectorXd> x_stdvec{x_vec, x_vec};
  stan::test::expect_ad(f, x_stdvec, sigma);
  stan::test::expect_ad(ff, x_stdvec, x_stdvec, sigma);
}
```

I think if we make the signatures use `require_vector_like_t<>` a good few of these can also support eigen matrix inputs

**Comment by SteveBronder** [comment]
Can you share any benchmarks you did? When @bbbales2 and I tried to fix these up we had a pretty big slowdown at small N sizes

File: stan/math/prim/fun/gp_dot_prod_cov.hpp (hunk 1 above)
Comment (t4c1) [reply stan/math/prim/fun/gp_dot_prod_cov.hpp:70]: > I'm surprised most of the time isn't being spent in dot_product(x[i], x[j])

Virtually all time is spent in dot product. That is why it is faster if the data is in cache. Loop optimizations probably suffer, but less transfers outweighs extra loops.

**Comment by t4c1** [comment]
Here is one benchmark I actually saved plot of: gp_10

> I think if we make the signatures use require_vector_like_t<> a good few of these can also support eigen matrix inputs

Adding support for more signatures and more tests is a separate thing and should be a separate PR.

**Comment by SteveBronder** [comment]
>> I think if we make the signatures use require_vector_like_t<> a good few of these can also support eigen matrix inputs

> Adding support for more signatures and more tests is a separate thing and should be a separate PR.

Yeah that's fine.

If you don't have time I can try to run benchmarks later, but we really need to check small N speed here. That looks like a 10% loss for anything below size 50

**Comment by t4c1** [comment]
I also reran this benchmark multiple times. The slowdown around 50 is just noise, although the slowdown below 5 is real.

**Comment by SteveBronder** [comment]
What am I calling wrong in the benchmark script below?

```
./benchmarks/benchmark.py gp_dot_prod_cov --multiplier 2 --max_size 8094 --repeats 10 --csv "gp_dot_prod_cov.csv"
```

It's giving me back 

```
$ ./benchmarks/benchmark.py gp_dot_prod_cov --multiplier 2 --max_size 8094 --repeats 10 --csv "gp_dot_prod_cov.csv"
make: 'test/expressions/stanc' is up to date.

make ./benchmarks/benchmark
g++ -march=native -mtune=native -std=c++1y -pthread -D_REENTRANT -Wno-sign-compare -Wno-ignored-attributes      -I lib/tbb_2020.3/include   -O3  -I . -I lib/eigen_3.3.9 -I lib/boost_1.75.0 -I lib/sundials_5.7.0/include -I lib/benchmark_1.5.1/googletest/googletest/include -I lib/benchmark_1.5.1/googletest/googletest -I lib/benchmark_1.5.1/include      -DBOOST_DISABLE_ASSERTS        -DGTEST_HAS_PTHREAD=0         -Wl,-L,"stan/origin/math/lib/tbb" -Wl,-rpath,"stan/origin/math/lib/tbb"     -L lib/benchmark_1.5.1/build/src   benchmarks/benchmark.cpp lib/benchmark_1.5.1/googletest/googletest/src/gtest-all.o lib/tbb/libtbb.so.2        -Wl,-L,"stan/origin/math/lib/tbb" -Wl,-rpath,"stan/origin/math/lib/tbb"   -lbenchmark -o benchmarks/benchmark

./benchmarks/benchmark --benchmark_repetitions=10 --benchmark_display_aggregates_only=true --benchmark_out=gp_dot_prod_cov.csv --benchmark_out_format=csv
Failed to match any benchmarks against regex: .
```

File: stan/math/prim/fun/gp_dot_prod_cov.hpp (hunk 1 above)
Comment (SteveBronder) [reply stan/math/prim/fun/gp_dot_prod_cov.hpp:70]: What if we just had an if that for x_size < block_size did the original and for > did the above version?

**Comment by t4c1** [comment]
Hmm, that is a bug in bench script. It was incorrectly identifying arrays as having many dimensions, so it skipped them by default. I am pushing a fix.

**Comment by t4c1** [comment]
Also I remembered that I did some changes that could affect performance after the benchmark I posted earlier. So I ran one benchmark and now it looks faster for all sizes. gp2

File: stan/math/prim/fun/gp_dot_prod_cov.hpp (hunk 1 above)
Comment (t4c1) [reply stan/math/prim/fun/gp_dot_prod_cov.hpp:70]: While I could add that, I think the extra code is not worth it just for ~5% speedup at tiny sizes. Also with the new bench ne code looks faster regardless of size.

**Comment by SteveBronder** [comment]
That's awesome! The code itself looks good let me run the benchmarks locally then I think this is good

**Comment by rok-cesnovar** [comment]
I was able to confirm Tadej's findings locally, was just looking at the code. If you say its good I can stop looking :)

**Comment by SteveBronder** [comment]
Cool cool I'll just check the code then

**Review by SteveBronder — CHANGES_REQUESTED** [review]
Did a first review, the rev versions look good, but I wrote some tests for the gp functions that use our testing framework and it looks like there is some odd behavior go `gp_periodic_cov` that's failing. They are in the branch below. Can you add simple tests like the ones I wrote below for the other GP functions to make sure we can catch these. Oddly, the failure is in the prim version  that takes in two vectors. Also the testing framework can't handle functions with 4 inputs so I just copy `l` to `p`

https://github.com/stan-dev/math/tree/review/gp-optims-tests

File: stan/math/prim/fun/gp_exp_quad_cov.hpp [hunk 2]
```diff
@@ -108,42 +85,17 @@ gp_exp_quad_cov(const std::vector<T_x1> &x1, const std::vector<T_x2> &x2,
...
-  using std::exp;
-  Eigen::Matrix<return_type_t<T_x1, T_x2, T_s>, Eigen::Dynamic, Eigen::Dynamic>
-      cov(x1.size(), x2.size());
-  for (size_t i = 0; i < x1.size(); ++i) {
-    for (size_t j = 0; j < x2.size(); ++j) {
-      cov(i, j) = sigma_sq * exp(-0.5 * (x1[i] - x2[j]).squaredNorm());
+  for (size_t ib = 0; ib < x1.size(); ib += block_size) {
+    for (size_t jb = 0; jb < x2.size(); jb += block_size) {
+      for (size_t j = jb; j < std::min(x2.size(), jb + block_size); ++j) {
+        for (size_t i = ib; i < std::min(x1.size(), ib + block_size); ++i) {
```
[the remark this answers is not in the document]
Comment (t4c1) [reply stan/math/prim/fun/gp_exp_quad_cov.hpp:93]: I am pretty sure the compiler can do that on its own. But I can do that if you find it more readable.

File: stan/math/prim/fun/gp_dot_prod_cov.hpp [hunk 3]
```diff
@@ -54,15 +54,21 @@ gp_dot_prod_cov(const std::vector<Eigen::Matrix<T_x, Eigen::Dynamic, 1>> &x,
...
+    for (size_t ib = jb; ib < x_size; ib += block_size) {
+      for (size_t j = jb; j < std::min(x_size, jb + block_size); ++j) {
+        cov.coeffRef(j, j) = sigma_sq + dot_self(x[j]);
+        for (size_t i = std::max(ib, j + 1);
+             i < std::min(x_size, ib + block_size); ++i) {
+          cov.coeffRef(j, i) = cov.coeffRef(i, j)
+              = sigma_sq + dot_product(x[i], x[j]);
+        }
+      }
     }
```
[the remark this answers is not in the document]
Comment (t4c1) [reply stan/math/prim/fun/gp_dot_prod_cov.hpp:69]: I doubt that can affect performance in any significant way, because it is not in the innermost loop (which is hidden in `dot_product` call).

File: stan/math/rev/fun/gp_exp_quad_cov.hpp [hunk 4]
```diff
@@ -208,11 +28,11 @@ class gp_exp_quad_cov_vari<T_x, double, T_l> : public vari {
...
  *   x is nan or infinite
  */
-template <typename T_x,
-          typename = require_arithmetic_t<typename scalar_type<T_x>::type>>
-inline Eigen::Matrix<var, -1, -1> gp_exp_quad_cov(const std::vector<T_x> &x,
-                                                  const var &sigma,
-                                                  const var &length_scale) {
+template <typename T_x, typename T_sigma, require_st_arithmetic<T_x>* = nullptr,
+          require_stan_scalar_t<T_sigma>* = nullptr>
+inline Eigen::Matrix<var, -1, -1> gp_exp_quad_cov(const std::vector<T_x>& x,
```
[the remark this answers is not in the document]
Comment (t4c1) [reply stan/math/rev/fun/gp_exp_quad_cov.hpp:33]: I don't think we ever put derivatives in the doc. I am not sure why would we start here.

**Comment by t4c1** [comment]
I don't know what is the issue with your tests. I find the ad tests quite hard to debug. Whatever it is, it does not seem to be from this PR, as the same test is failing on develop. Possibly something is wrong with the test itself.

File: stan/math/rev/fun/gp_exp_quad_cov.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/rev/fun/gp_exp_quad_cov.hpp:33]: If you look in the docs for rev there's a bunch of functions that have docs giving the derivatives. That or if the derivatives come from a paper they are cited there

**Comment by SteveBronder** [comment]
It could be there's something wrong with the test since the testing framework is not made to handle functions with 4 arguments. 

> I don't know what is the issue with your tests. I find the ad tests quite hard to debug.

Anything in particular? I've thought about cleanup up a few things there, like for instance I think it would be be nice to have better  debugging style stuff to happen when a test fails like printing out the failing matrices. gtest let's you wait to fail instead of immediately failing which does allow you to collect more debug info before calling the test a failure

**Comment by t4c1** [comment]
> Anything in particular?

It is virtually impossible to set a breakpoint so that it breaks only when one (lets say rev) overload is called. Also there is huge amount of code and lambdas, so for me it is hard to follow what certain part is responsible for. Converting arguments into std vectors and back also does not help clarity. Better error messages, such as at which index the error is, would also help.

Now that you mention it, I kinda want to refactor it myself. Possibly merging it with distribution testing framework.

**Comment by SteveBronder** [comment]
> Now that you mention it, I kinda want to refactor it myself. Possibly merging it with distribution testing framework.

Let's open up an issue shopping some ideas

Also sorry for the delay in reviewing this. I just want to check the periodic, rest of the code looks good

**Review by SteveBronder — APPROVED** [review]
checked out periodic locally and lgtm!
